---
title: "A very short intro to R"
author: "Francesco Montinaro"
date: "10/01/2022"
output:
  ioslides_presentation:
        #logo: R_logo.svg.png
incremental: yes

---



<style type="text/css">
body p {
display: block;
  color: #36383b;
  font-size: 20px;
}
ul {
  display: block;
  font-size: 20px;
  margin-top: 0;
  margin-bottom: 0;
  margin-left: 0;
  margin-right: 0;
  font-weight: bold;
  color:#36383b;
}
h2 { font-size: 35px;}
.title-slide hgroup h2 {
  font-size: 25px;
  letter-spacing: 20;
}

</style>


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE)
```

<!-- ## A very brief history of R -->

<!-- - R is an implementation of S language, created in 1976. -->

<!-- ## A very brief history of R -->

<!-- - R is an implementation of S language, created in 1976. -->
<!-- - A modified version of S (S-PLUS) was developed in 1993. -->

## A very brief history of R


<!-- ![](R_logo.svg.png){width=20%} -->

>- R is an implementation of S language, created in 1976.
>- A modified version of S (S-PLUS) was developed in 1993.
>- Developed in 1991 by Ross Ihaka and Robert Gentleman and became free in 1995

<br>
<br>
<br>
<br>
<br>

```{r, echo=FALSE, fig.align='right', fig.margin=T,out.width = 100, out.height = 100}
knitr::include_graphics("./img/R_logo.svg.png",dpi=200)
```

## Why R?

>- Open source
>- Reasonable fast 
>- Operators for array (eg. matrix) analysis
>- Active development (packages)
>- Active user community
>- Several packages for data visualization
>- Integration with distributed version control system

## But R ...

>- It's not user friendly
>- Has no commercial support
>- It's limited by RAM amount
>- Steep learning curve

## How to use R

R can be downloaded [here](https://www.r-project.org/).

One of the most used ways to use R is through the free and open source (but there is also commercial version)
Rstudio Desktop, which helps to control all the different entities in R.


Rstudio can be downloaded [here](https://rstudio.com/).

## Rstudio: First steps

```{r, echo=FALSE, fig.align='center', fig.margin=T,out.height="100%",out.width="100%"}
knitr::include_graphics("./img/Rstudio.png",dpi=1400)
```

## Rstudio: Set the working directory

- Create a folder with name "esercitazione"
- Copy the file "gene_expression_data.txt" in the folder "esercitazione"
- Open Rstudio
- Go to Session --> Set Working directory --> Choose Directory...
- Click Open and confirm


## Rstudio: Set the working directory

- Go to File --> New File --> R script
- Go to File --> Save --> Put in the esercitazione folder with name es1.R
- In the "Source Panel" type **3+9** and press Run
- On a new line type **45/8** and press Run
- You can use the shortcut ***ctrl + Enter*** (Windows/Linux) or ***CMD + Enter*** to run a single line 
- ***ctrl + shift + Enter *** (Windows/Linux) or ***CMD + shift + Enter*** run all the code

## R operations:

```{r, include=T,echo=T}
75+4
41-9
99/12
17*6
```

## R operations:

```{r, include=T,echo=T}
9^3
27%%3
44%%7
```
## R logical operations:
<br>
<br>
```{r, echo=T, results='hide'}
10e4==100000
24!=34
15>20
30<21
```

## R logical operations:
<br>
```{r, echo=T, include=T}
10e4==100000
24!=34
```

## R logical operations:
<br>
```{r, echo=T, include=T}
15>20
30<21
```




## R characters:
R is also able to manage character variables which must be surronded by "" or ' '

<br>
```{r echo=T, include=T, paged.print=TRUE}
"Hello world!"

"I am a string"
```


## Vectors, objects and assignment
### **Vectors:** ###
- Numeric vectors

```{r, echo=T, include=T}
c(10, 30, 50,12)
```
Note the *c()* function which means 'concatenate'

- Character vectors

```{r, echo=T, include=T}
c("I", "am", "a","string vector")
```

## Assign variable to objects

*Variables* can be assigned to object using the "<-" or "=" symbol.

```{r, echo=T, include=T}
marks <- c(26, 30, 28,19)

basket <- c("peach", "orange", "cherries", "pineapple")
```


The content of any object can be shown just typing its name.

```{r, echo=T, include=T}
marks

basket
```

## Classes : numeric and characters

We can check the class of any object with the function class(objectname)

```{r, echo=T, include=T}
class(marks)

class(basket)
```


## Classes: Factors

**Factors** are variables in R which take on a limited number of different values;
They are therefore **Categorical**, referring to categories that are called **levels** in R syntax.

```{r, echo=T, include=T}
myBiologicalType <- c("coding","UTR","intron","coding",
                      "UTR","regulatory","UTR")
myBiologicalType <- as.factor(myBiologicalType)
levels(myBiologicalType)
``` 


## Classes: Factors

Although it's not initially obvious, **factors** are very important. Consider, for example a vector of 100,000 entries summarizing the sex of indivuduals for a given species.
```{r, echo=T, include=T,message=FALSE}
library(pryr)
sexes <- sample(c("male","female"),size = 100000,replace = T)
length(sexes)
object_size(sexes)
object_size(as.factor(sexes))
```

## Working on vectors

In R, objects can be modified in many ways:

let's create two vectors:

```{r, echo=T, include=T,message=FALSE}
vec1 <- 1:10 # What am I doing here?

vec2 <- seq(from=100,to = 1000,by = 100) # And here?
```


## Working on vectors

The length of the vector can be retrieved through the *length()* function: 

```{r, echo=T, include=T,message=FALSE}
length (vec1)

length(vec2)
```

## Working on vectors: manipulation

we can subset vectors indicating the index in square bracket


```{r, echo=T, include=T,message=FALSE}
vec2 #show the full vector

vec2[3] # show the third element of the vector

vec2[8] # show the eighth element of a vector

```
## Working on vectors :manipulation

we can subset vectors indicating the index in square bracket


```{r, echo=T, include=T,message=FALSE}
vec2 #show the full vector

vec2[3:6] # show FROM the third element TO the sixth element of the vector

vec2[c(2,5,9)] # show the 2nd,5th and 9th element

```

## Working on vectors :manipulation

We can replace values in vectors using this syntax:


```{r, echo=T, include=T,message=FALSE}
vec2 #show the full vector
vec2[3] # show the third element
vec2[3] <- 450  # replace the 3rd elemnt with 450
vec2[3]
```


## Working on vectors: operations

When working with vectors it is possible to apply an operation to all the variables at once:

```{r, echo=T, include=T,message=FALSE}
vec2
vec2/2  # replace the 3rd elemnt with 450
vec2-30
```
## Working on vectors: operations

When working with vectors it is possible to apply an operation to all the variables at once:

```{r, echo=T, include=T,message=FALSE}
vec2
vec2*2
vec2+50
```

## Working on vectors: operations

We can also perform operations between vectors:

```{r, echo=T, include=T,message=FALSE}
vec2 #show vec2

vec1 # show vec1

vec1*vec2 # Multiply vec1 and vec2


```

## Working on vectors: operations

We can also perform operations between vectors:

```{r, echo=T, include=T,message=FALSE}
vec2 #show vec2

vec1 # show vec1


vec1/vec2 # Divide vec2 and vec1

```

## Working on vectors: operations


We can also perform operations between vectors:

```{r, echo=T, include=T,message=FALSE}
vec2 #show vec2

vec1 # show vec1

vec1+vec2 # sum vec1 and vec2


```

## Working on vectors: operations


We can also perform operations between vectors:

```{r, echo=T, include=T,message=FALSE}
vec2 #show vec2

vec1 # show vec1

vec1-vec2 # Subtract vec2 to vec1

```

## Working on vectors: operations

<center>**WARNING!**</center>

```{r, echo=FALSE, fig.align='center', fig.margin=T}
knitr::include_graphics("./img/warning.png",dpi=300)
```

When performing operations between vectors, the elements of the shorter one are ***recycled***.

```{r, echo=T, include=T,message=T}
v1 <- 1:5
v2 <- c(2,3)

v1*v2 # What is going on here?

```


## Working on vectors: recycling

<!-- #<center>**WARNING!**</center> -->
<br>

```{r, echo=FALSE, fig.align='center', fig.margin=T}
knitr::include_graphics("./img/recycling1.png",dpi = 100)
```

## Working on vectors: recycling

<!-- #<center>**WARNING!**</center> -->
<br>

```{r, echo=FALSE, fig.align='center', fig.margin=T}
knitr::include_graphics("./img/recycling2.png",dpi = 100)
```

## Working on vectors: recycling

<!-- #<center>**WARNING!**</center> -->
<br>

```{r, echo=FALSE, fig.align='center', fig.margin=T}
knitr::include_graphics("./img/recycling3.png",dpi = 100)
```

## Data visualization : Scatterplot

```{r, echo=T, include=T,message=T}
x <- 1:10 # create a vector from 1 to 10
y <- 51:60 # create a vectro from 51:60
plot(x,y) #plot x and y
```

## Data visualization : Scatterplot

```{r, echo=T, include=T,message=T}
x <- 1:10 # create a vector from 1 to 10
y <- 51:60 # create a vector from 51:60
plot(x,y) # plot x and y
lines(x,y,col="red") #add a red line 
```

## Data visualization : histogram

```{r, echo=T, include=T,message=T}
x <- rnorm(1000,mean=10) # create random normal 
                         # distribution of 1,000 samples with mean=10

hist(x)

```

## Matrices and data.frames

```{r, echo=F, include=T,message=T}

matrix(round(runif(40,min=1,max=100)),nrow = 5)
```
>- **Matrices** are R objects in which data are organised in a 2-dimensional array.
>- they can contain only elements of the same types (eg. only numbers, only characters, etc.).
>- A matrix is created through the *matrix()* function. 

## Matrices

```{r, echo=T, include=T,message=T}
m=matrix(1:100,nrow=10)
m
```

## Matrices


```{r, echo=T, include=T,message=T}
m <- matrix(1:100,nrow=10,byrow = T)
m
```

## Matrices

```{r, echo=T, include=T,message=T}
class(m) # shows the class of the object m

is.matrix(m) # shows TRUE if m is a matrix, FALSE otherwise

dim(m) # shows the number of rows and column of object m

nrow(m) # shows the number of rows ob object m

ncol(m) # shows the number of columns of object m

```

## Matrices: data extraction

```{r, echo=T, include=T,message=T}
m[3,] # extract the third row 
m[,1] # extract the first column
```


## Matrices: data extraction

```{r, echo=T, include=T,message=T}
# extract the values of the 2nd and third row 
# in the 5th, 6th and 9th column
m[2:3,c(5,6,9)] 
```

```{r, echo=T, include=T,message=T}
m[6,7] # extract the value of the 6th row and senth column
```

## Dataframes
An important R class is *dataframe*, which is characterised by high flexibility. In fact despite *matrices*, dataframes can contain multiple types of data  

```{r, echo=T, include=T,message=T}
gene <-  c("ADGRB2", "TARBP1", "SLC22A24", "SLC15A5", 
           "LOC117975386", "C17H17orf99", "TJP3", "TRIB3")
expression_in_cases <-c(100, 30, 35, 66,90, 200,12,8) 
expression_in_controls <-c(10, 45, 55, 70,110, 80,25,85) 
```
We can create a data.frame using the **data.frame()** function:

```{r, echo=T, include=T,message=T}
df <- data.frame(gene,expression_in_cases,expression_in_controls)
```
dataframes allow for data extraction using the same notations we saw before for *matrices*
So, what the code needed to show only the first two rows of the data.frame?

## Dataframes
An important R class is *dataframe*, which is characterised by high flexibility. In fact despite *matrices*, dataframes can contain multiple types of data  

```{r, echo=T, include=T,message=T}
gene <-  c("ADGRB2", "TARBP1", "SLC22A24", "SLC15A5", "LOC117975386", "C17H17orf99", "TJP3", "TRIB3")
expression_in_cases <-c(100, 30, 35, 66,90, 200,12,8) 
expression_in_controls <-c(10, 45, 55, 70,110, 80,25,85) 
```
We can create a data.frame using the **data.frame()** function:

```{r, echo=T, include=T,message=T}
df <- data.frame(gene,expression_in_cases,expression_in_controls)
```
So, what the code needed to show only the first two rows of the data.frame?

```{r, echo=T, include=T,message=T}
df[1:2,]
```

## Dataframes: data extraction
```{r, echo=T, include=T,message=T}
df[1:2,]
```

dataframes also allow to select specific columns usig the **$** notations

```{r, echo=T, include=T,message=T}
df$gene #show the column "gene"
df$expression_in_cases # show the column "expression_in_cases" 
```

## Dataframes: transformation
A *dataframe* can be converted in *matrix* using the function **as.matrix()**

```{r, echo=T, include=T,message=T}
m <- as.matrix(df)
class(m)
```
We can convert a matrix to a datafram using *as.dataframe()*
```{r, echo=T, include=T,message=T}
df <- as.data.frame(m)
class(df)
```
## Dataframes: transformation
We can also explore the structure of the object using **str()**

```{r, echo=T, include=T,message=T}
str(m) # show the structure of the object m
str(df) # show the structure of the object df
```

## R data structures: Overview
```{r, echo=F, include=T,message=T}
object <- c("vector","matrix","dataframe")
modes <-  c("numeric, character, complex, logical","numeric, character, complex, logical","numeric, character, complex, logical")
Allow_multiple_modes=c("No","No","yes")
dft <- data.frame(object,modes,Allow_multiple_modes)
```

Here is a brief overview of **some** fundamental data structure in **R**
```{r, echo=F, include=T,message=T}
library(knitr)
knitr::kable(dft)
```


## R: Challenge yourself ##

<br>

 1. Open Rstudio
 2. Type ```set.seed(123)```
 3. Create a vector **m** of length 20, composed by numbers drawn by a normal distribution with mean=0 and sd=1
 4. Create a vector **n** with the first 5 elements from the vector **m**
 5. Plot an histogram of **m**

 
## R: Challenge yourself ##

<br>

 6. Create a matrix **mat** composed by 10 rows and 10 columns with numbers from 201...300
 7. Replace the value located at the second row, third column with 84
 8. Plot a *scatterplot* of the second column of mat in x axis and the and fifth column of mat on the y axis
 9. Convert the matrix **mat** into a dataframe **df**
 10. Show the column "V2" of the dataframe *df* using the **$** notation
 

## R: Challenge yourself ##

```{r, echo=T, include=T,message=T}
set.seed(123) #2. Type ```set.seed(123)```
m=rnorm(20,mean=0,sd=1) # 3. Create a vector **m** of length 20, composed by numbers drawn by a normal distribution with mean=0 and sd=1
n=m[1:5] # Create a vector **n** with the first 5 elements from the vector **m**
hist(m) # 5. Plot an histogram of **m**

```

## R: Challenge yourself ##

```{r, echo=T, include=T,message=T}
# 6. Create a matrix **mat** composed 
#by 10 rows and 10 columns with numbers from 201...300
mat <- matrix (data = 201:300,nrow=10,ncol=10) 
# 7. Replace the value located 
# at the second row, third column with 84
mat[2,3] <- 84  
```

## R: Challenge yourself ##

```{r, echo=T, include=T,message=T}
# 8. Plot a *scatterplot* of the second column of
# mat in x axis and the and fifth column of mat on the y axis
plot(mat[,2],mat[,5]) 
```


## R: Challenge yourself ##

```{r, echo=T, include=T,message=T}
#9. Convert the matrix **mat** into a dataframe **df**
df=as.data.frame(mat) 
#10. Show the column "V2" of the dataframe *df* using the **$** notation
df$V2  
```
## R: working with real datasets

We are now ready to work on a real problem. 

Let's assume you got 1000 euros to do a pilot study on the genetic basis of drug addiction.

I want to be quick and efficient, and I need to prioritise my research towards genes that are ACTUALLY relate to drug addiction.

After hours and hours of literature review I found a list of 12 candidate genes, and their expression at 54 different tissues. They are saved on the file "gene_expression_data.txt"

## R: working with real datasets
1. Open Rstudio
2. Set the working directory to "esercitazione"
3. Let's read the "gene_expression_data.txt" file in R
```{r, echo=T, include=T,message=T}
gene_expression <- read.table(file = "./data/gene_expression_data.txt",
                              header = T,sep = "\t",check.names = F)
```

## R: working with real datasets
4. Let's have a look at the dataset
```{r, echo=T, include=T,message=T}
gene_expression[1:5,1:5]
```
## R: working with real datasets
5. Show the list of genes in the dataset 

## R: working with real datasets
6. Show the list of genes in the dataset 

```{r, echo=T, include=T,message=T}
gene_expression$Description
```

## R: working with real datasets
6. Show the list of tissues in the dataset 

```{r, echo=T, include=T,message=T}
colnames(gene_expression)[-c(1:2)]
```

## R: working with real datasets
One way to decide to which gene put more effort (and money) I will look their expression in different brain tissues using 
the **barplot()**. I will focus on the following tissues

<br>
<br>
<br>

- Brain - Cerebellum
- Brain - Cortex
- Brain - Hippocampus
- Brain - Hypothalamus

## R: working with real datasets
Brain - Cerebellum
```{r, echo=T, include=T,message=T, fig.height=3.5}
barplot(gene_expression$`Brain - Cerebellum`,
        names.arg = gene_expression$Description,las=2, 
        col="blue",cex.names = .8,ylab="transcript per million")
```

## R: working with real datasets
Brain - Cortex
```{r, echo=T, include=T,message=T,fig.height=3.5}
barplot(gene_expression$`Brain - Cortex`,
        names.arg = gene_expression$Description,las=2, 
        col="green",cex.names = .8,ylab="transcript per million")
```


## R: working with real datasets
Brain - Hippocampus
```{r, echo=T, include=T,message=T,fig.height=3.5}
barplot(gene_expression$`Brain - Hippocampus`,
        names.arg = gene_expression$Description,las=2, 
        col="red",cex.names = .8,ylab="transcript per million")
```

## R: working with real datasets
Brain - Hypothalamus
```{r, echo=T, include=T,message=T,fig.height=3.5}
barplot(gene_expression$`Brain - Hypothalamus`,
        names.arg = gene_expression$Description,las=2, 
        col="purple",cex.names = .8,ylab="transcript per million")
```

## R: Wrap it all- Which gene is a good candidate to study?

<br>
<br>
```{r, echo=T,results='hide', include=T,message=F,fig.height=3.5}
par(mfrow=c(2,2),mar=c(6,2.5,1.5,1))

barplot(gene_expression$`Brain - Cerebellum`,
        names.arg = gene_expression$Description,las=2, 
        col="blue",main="Cerebellum")
barplot(gene_expression$`Brain - Cortex`,
        names.arg = gene_expression$Description,las=2, 
        col="green",main="Cortex")
barplot(gene_expression$`Brain - Hippocampus`,
        names.arg = gene_expression$Description,las=2, 
        col="brown",main="Hippocampus")
barplot(gene_expression$`Brain - Hypothalamus`,
        names.arg = gene_expression$Description,las=2, 
        col="purple",main="Hypotalamus")
```



## R: Wrap it all- Which gene is a good candidate to study?

```{r, echo=F, include=T,message=T,fig.height=3.5}
par(mfrow=c(2,2),mar=c(6,2.5,1,1))

barplot(gene_expression$`Brain - Cerebellum`,
        names.arg = gene_expression$Description,las=2, col="blue",main="Cerebellum")
barplot(gene_expression$`Brain - Cortex`,
        names.arg = gene_expression$Description,las=2, col="green",main="Cortex")
barplot(gene_expression$`Brain - Hippocampus`,
        names.arg = gene_expression$Description,las=2, col="brown",main="Hippocampus")
barplot(gene_expression$`Brain - Hypothalamus`,
        names.arg = gene_expression$Description,las=2, col="purple",main="Hypotalamus")
```